beacon_node/beacon_chain/src/block_verification.rs

```diff
@@ -1454,58 +1454,49 @@ impl<T: BeaconChainTypes> ExecutionPendingBlock<T> {
-                    chain.store.load_hot_state_summary(&state_root)?.is_some();
-
-                let state_batch = if state_already_exists {
-                    // If the state exists, it could be temporary or permanent, but in neither case
-                    // should we rewrite it or store a new temporary flag for it. We *will* stage
-                    // the temporary flag for deletion because it's OK to double-delete the flag,
-                    // and we don't mind if another thread gets there first.
-                    vec![]
+            let state_root =
+                if parent.beacon_block.slot() == state.slot() {
+                    // If it happens that `pre_state` has *not* already been advanced forward a single
+                    // slot, then there is no need to compute the state root for this
+                    // `per_slot_processing` call since that state root is already stored in the parent
+                    // block.
+                    parent.beacon_block.state_root()
                 } else {
-                    vec![
-                        // TODO(hdiff): No longer relevant under HDiff hot
-                        if state.slot() % T::EthSpec::slots_per_epoch() == 0 {
-                            StoreOp::PutState(state_root, &state)
-                        } else {
-                            StoreOp::PutStateSummary(
-                                state_root,
-                                // Doing this computation here feels spagetti
-                                HotStateSummary::new(
-                                    &state_root,
-                                    &state,
-                                    &chain.store.hierarchy_hot,
-                                )?,
-                            )
-                        },
-                        StoreOp::PutStateTemporaryFlag(state_root),
-                    ]
-                };
-                chain
-                    .store
-                    .do_atomically_with_block_and_blobs_cache(state_batch)?;
-                drop(txn_lock);
+                    // This is a new state we've reached, so stage it for storage in the DB.
+                    // Computing the state root here is time-equivalent to computing it during slot
+                    // processing, but we get early access to it.
+                    let state_root = state.update_tree_hash_cache()?;
+
+                    // Store the state immediately, marking it as temporary, and staging the deletion
+                    // of its temporary status as part of the larger atomic operation.
+                    // TODO(hdiff): Is it necessary to do this read tx now? Also why is it necessary to
+                    // check that the summary exists at all? Are double writes common? Can this txn
+                    // lock deadlock with the `do_atomically` call?
+                    let txn_lock = chain.store.hot_db.begin_rw_transaction();
```

Comment:
The reason for this lock is a race with the write in `BeaconChain::import_block`:

If we were to write a non-temporary state in `import_block` in between setting `state_already_exists` (`false`) and the write of the temporary state in this function, we can corrupt the DB:

- There is a state that is _not_ temporary (required by some fully-imported block),
- But it is marked temporary due to the race condition here
- Temporary states risk being deleted by pruning -> invalid DB due to deletion of canonical state.

The lock prevents this case by preventing the interleaving of the read in this function with the write in the import function. However this is a bad abstraction forced upon us by LevelDB which lacks proper ACID transactions. If we move away from LevelDB eventually we can maybe have proper transactions, see:

- https://github.com/sigp/lighthouse/pull/4718